pyccel/parser/syntactic.py

```diff
@@ -1243,20 +1272,34 @@ def _visit_GeneratorExp(self, stmt):
             body = Assign(lhs, body)
 
         body.set_current_ast(parent)
-        indices = []
-        generators = list(generators)
-        while len(generators) > 0:
-            indices.append(generators[-1].target)
+
+        for loop, condition in zip(generators, conditions):
+            if condition:
+                loop.insert2body(condition)
+
+        if generators[-1].body.body:
+            generators[-1].body.body[0].blocks[0].body.insert2body(body)
```

Comment:
Can we add some intermediate variables or comments to make this easier to follow?